src/encoder.js

```diff
@@ -270,41 +271,100 @@ export default class Encoder extends Extendable {
+    while (Object.keys(this.placeholderMap).length > 0) {
+      let placeholderMap = this.placeholderMap;
+      let replacements = [];
+      let enabledExtensions = extensions.slice(0);
+
+      // run the logic for all extensions
+      while (enabledExtensions.length > 0) {
+        let extensionPoint = enabledExtensions.pop();
+        let extension = this.extensions[extensionPoint];
+        let applyRecursively = extension.shouldApplyRecursively != null && extension.shouldApplyRecursively();
+
+        if (extensionPoint in placeholderMap) {
+          let placeholders = placeholderMap[extensionPoint];
+          let originalSize = placeholders.length;
+          for (let i = 0; i < originalSize; ++i) {
+            let placeholder = placeholders[i];
+            this.placeholderMap = {};
+            let extTarget = [];
+            if (extension.shouldSerialise == null || extension.shouldSerialise(placeholder.value)) {
+              if (extensionPoint < 8) {
+                extTarget.push(tags.EXTENSION3_BASE | extensionPoint);
+              } else {
+                extTarget.push(tags.EXTENSION_);
+                encodeUInt(extensionPoint, extTarget);
+              }
+              let furtherEnabledExtensions = applyRecursively
               ? [...enabledExtensions, extensionPoint]
               : enabledExtensions;
-            this.encodeValue(extension.serialise(placeholder.value), extTarget, furtherEnabledExtensions);
-          } else {
-            this.encodeValue(placeholder.value, extTarget, enabledExtensions);
+              this.encodeValue(extension.serialise(placeholder.value), extTarget, furtherEnabledExtensions);
+            } else {
+              this.encodeValue(placeholder.value, extTarget, enabledExtensions);
+            }
+            replacements.push({ index: placeholder.location, data: extTarget, placeholders: this.placeholderMap });
           }
-          target.splice(target.indexOf(placeholder), 1, ...extTarget);
         }
       }
 
-      if (typeof extension.memo === 'function') {
-        target.unshift(...this.encodeValue(extension.memo(), [], enabledExtensions));
+      this.placeholderMap = {};
+
+      let returnData = [];
+      let lastIndex = 0;
+
+      // restructure data into new array
+      for (let replacement of replacements) {
+        if (replacement.index < lastIndex) {
+          throw new Error('Corrupt non-linear state');
+        } else if (replacement.index > lastIndex) {
+          returnData.push(...target.slice(lastIndex, replacement.index)); // skips placeholder byte
+        }
+        for (let extensionPoint in replacement.placeholders) {
+          if (!(extensionPoint in this.placeholderMap)) {
+            this.placeholderMap[extensionPoint] = [];
+          }
+          let placeholders = replacement.placeholders[extensionPoint];
+          for (let placeholder of placeholders) {
```

Comment:
I think we avoid using `for-of` in this project due to its dependency on symbols. I may be mistaken though.